File: samples/java/Bingo/Client/src/main/java/systems/zlink/samples/bingo/client/configuration/SampleTopology.java
package systems.zlink.samples.bingo.client.configuration;

import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

public record SampleTopology(String sessionAStreamEndpoint, String sessionBStreamEndpoint) {
    public static SampleTopology load(String[] args) {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: Client --config <path>");
        }
        Properties properties = new Properties();
        try (Reader reader = Files.newBufferedReader(Path.of(args[1]))) {
            properties.load(reader);
        } catch (Exception error) {
            throw new IllegalStateException("Could not load Bingo client config.", error);
        }
        return new SampleTopology(
                required(properties, "sessionAStreamEndpoint"),
                required(properties, "sessionBStreamEndpoint"));
    }

    private static String required(Properties properties, String name) {
        String value = properties.getProperty(name);
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("Missing Bingo client config: " + name);
        }
        return value;
    }
}

File: samples/kotlin/Bingo/Client/src/main/kotlin/systems/zlink/samples/kotlin/bingo/client/configuration/SampleTopology.kt
package systems.zlink.samples.kotlin.bingo.client.configuration

import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties

data class SampleTopology(val sessionAStreamEndpoint: String, val sessionBStreamEndpoint: String) {
    companion object {
        fun load(args: Array<String>): SampleTopology {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: Client --config <path>"
            }
            val properties =
                Properties().also { values ->
                    Files.newBufferedReader(Path.of(args[1])).use(values::load)
                }
            return SampleTopology(
                required(properties, "sample.session-a-stream-endpoint"),
                required(properties, "sample.session-b-stream-endpoint"),
            )
        }

        private fun required(properties: Properties, name: String): String =
            requireNotNull(properties.getProperty(name)?.takeIf(String::isNotBlank)) {
                "$name is required"
            }
    }
}
